go/tasks/plugins/k8s/sagemaker/utils.go

```diff
@@ -0,0 +1,218 @@
+	case sagemakerSpec.HyperparameterTuningObjective_MAXIMIZE:
+		return MaximizeSageMakerAPIHyperParameterTuningJobObjectiveType
+	}
+	return MinimizeSageMakerAPIHyperParameterTuningJobObjectiveType
+}
+
+func getAPITrainingInputMode(trainingInputMode sagemakerSpec.InputMode) commonv1.TrainingInputMode {
+	switch trainingInputMode {
+	case sagemakerSpec.InputMode_FILE:
+		return FileSageMakerAPITrainingInputMode
+	case sagemakerSpec.InputMode_PIPE:
+		return PipeSageMakerAPITrainingInputMode
+	}
+	return FileSageMakerAPITrainingInputMode
+}
+
+func getAPITrainingJobEarlyStoppingType(
+	earlyStoppingType sagemakerSpec.HPOJobConfig_TrainingJobEarlyStoppingType) commonv1.TrainingJobEarlyStoppingType {
+
+	switch earlyStoppingType {
+	case sagemakerSpec.HPOJobConfig_OFF:
+		return OffSageMakerAPITrainingJobEarlyStoppingType
+	case sagemakerSpec.HPOJobConfig_AUTO:
+		return AutoSageMakerAPITrainingJobEarlyStoppingType
+	}
+	return OffSageMakerAPITrainingJobEarlyStoppingType
+}
+
+func getAPIAlgorithmName(name sagemakerSpec.AlgorithmName) string {
+	switch name {
+	case sagemakerSpec.AlgorithmName_CUSTOM:
+		return CustomSageMakerAPIAlgorithmName
+	case sagemakerSpec.AlgorithmName_XGBOOST:
+		return XgboostSageMakerAPIAlgorithmName
+	}
+	return CustomSageMakerAPIAlgorithmName
+}
+
+func getAllVersions(cfg *config.Config, algName, region string) []string {
+	allVers := make([]string, len(cfg.AlgorithmPrebuiltImages[algName][region]))
+	i := 0
+	for k := range cfg.AlgorithmPrebuiltImages[algName][region] {
+		allVers[i] = k
+		i++
+	}
+	return allVers
+}
+
+func convertRawVersToSemVers(raw []string) ([]*semver.Version, error) {
+
+	vs := make([]*semver.Version, len(raw))
+	for i, r := range raw {
+		v, err := semver.NewVersion(r)
+		if err != nil {
+			return nil, errors.Errorf("Failed to convert version string [%v] to semantic version [%v]", r, v)
+		}
+		vs[i] = v
+	}
+	return vs, nil
+}
+
+func getLatestSemVer(cfg *config.Config, algName, region string) (string, error) {
+	allVersionsRaw := getAllVersions(cfg, algName, region)
+	allSemVers, err := convertRawVersToSemVers(allVersionsRaw)
+	if err != nil {
+		return "", errors.Wrapf(err, "Failed to get the semantic versions of algorithm:region [%v:%v]", algName, region)
+	}
+	sort.Sort(semver.Collection(allSemVers))
+	// https://play.golang.org/p/bd7XndTKwq5
+	return allSemVers[len(allSemVers)-1].String(), nil
+}
+
+func getTrainingImage(job *sagemakerSpec.TrainingJob) (string, error) {
+	cfg := config.GetSagemakerConfig()
+	var err error
+	if specifiedAlg := job.GetAlgorithmSpecification().GetAlgorithmName(); specifiedAlg != sagemakerSpec.AlgorithmName_CUSTOM {
+		// Built-in algorithm mode
+		apiAlgorithmName := getAPIAlgorithmName(specifiedAlg)
+
+		// Getting the version
+		ver := job.GetAlgorithmSpecification().GetAlgorithmVersion()
+		if ver == "" {
+			// user didn't specify a version -> use the latest
+			ver, err = getLatestSemVer(cfg, apiAlgorithmName, cfg.Region)
```

Comment:
this is interesting, this introduces potential non-determinism to the plugin execution? I guess it is fine, but good to note

Reply:
ah good point...

Reply:
comment?